pkg/endpoint/bpf.go

```diff
@@ -962,6 +1014,58 @@ func (e *Endpoint) applyPolicyMapChanges(regenContext *regenerationContext, hasN
 	// returns changes that need to be applied to the Endpoint's bpf policy map.
 	closer, changes := e.desiredPolicy.ConsumeMapChanges()
 	defer closer()
+	changeSize := changes.Size()
+	// If the desiredPolicy will be larger than the BPF maximum after
+	// the changes we need to reset the consumed changes and either
+	// lockdown or return an inevitable error (which would happen on
+	// insertion if we proceed).
+	if e.desiredPolicy.GetPolicyMap().Len()+changeSize > policymap.MaxEntries {
+		// Revert the map changes that were just consumed.
+		e.desiredPolicy.RevertChanges(changes)
+		// We only need to go through the mechanics of
+		// lockdown once.
+		if !e.lockdown {
+			// Avoid an insertion error further down and return an
+			// error now.
+			if !option.Config.EnableEndpointLockdownOnPolicyOverflow {
+				// Do not return ErrPolicyEntryMaxExceeded as it
+				// implies that option.Config.EnableEndpointLockdownOnPolicyOverflow
+				// is "true".
+				return fmt.Errorf("policy map exceeds the map max entries limit")
+			}
+
+			e.getLogger().WithFields(logrus.Fields{
+				logfields.EndpointID: e.ID,
+			}).Warnf("The policy map exceeds the max entries limit, %s is enabled, locking the endpoint down.",
+				option.EnableEndpointLockdownOnPolicyOverflow)
+
+			if err := e.endpointPolicyLockdown(); err != nil {
+				e.getLogger().WithFields(logrus.Fields{
+					logfields.EndpointID: e.ID,
+				}).WithError(err).Error("Failed to lockdown endpoint:" +
+					"Consider quarantining or shutting down this node.")
+				return err
+			}
+			e.lockdown = true
+		}
+		e.updatePolicyMapPressureMetric(float64(e.desiredPolicy.GetPolicyMap().Len() + changeSize))
+		return ErrPolicyEntryMaxExceeded
```

Comment:
This will cause regeneration of this endpoint to permanently fail. That is probably not desired; there may be code that waits for all endpoints to be regenerated.

Rather, we will need to find a way to cleanly transition an endpoint in to lockdown mode that does not otherwise block regeneration.

Reply:
> This will cause regeneration of this endpoint to permanently fail. That is probably not desired; there may be code that waits for all endpoints to be regenerated.

Good catch.

>Rather, we will need to find a way to cleanly transition an endpoint in to lockdown mode that does not otherwise block regeneration.

I updated `runPreCompilationSteps` to swallow the error. No code in `runPreCompilationSteps` will be affected thereafter. I also updated `regenerateBPF` to avoid syncing `desiredPolicy` into `realizedPolicy` if the endpoint is in lockdown. Both those changes together will allow regeneration to proceed without affecting the lockdown state.